### ISSUE TITLE
basset有earlystop，用stagnant + epoch_best间接表示，这边需要做对应调整

### ISSUE THREAD

**Issue by sybwjdnr** [gao-lab/VCNN#9]
https://github.com/davek44/Basset/blob/8b672320c348c6610ad79015457e0afa1c18e456/src/basset_train.lua

while epoch <= opt.max_epochs and epoch - epoch_best <= opt.stagnant_t do

……
-- update best
    if valid_acc_avg > acc_best then
        io.write(" best!")
        acc_best = valid_acc_avg
        epoch_best = epoch

        -- save best
        torch.save(string.format('%s_best.th' % opt.save), convnet)
    end

1. monitor改为validation的accuracy val_accuracy
2. 确认“cmd:option('-stagnant_t', 10, 'Allowed epochs with stagnant validation loss')” 这样的话，keras这边的early stop的patience要写多少。lua这边是<=，也就是说，epoch_best是50的时候，如果你跑到了第60个epoch，那么还是会进入循环，进行梯度更新，评估val等，此时还是有希望更新epoch_best，导致下一轮epoch_best和epoch的差值又从1开始计算。只有在这一轮里epoch_best不更新，下一次循环开头才会中止。keras这边的earlystop要确认一下它的模式要如何配置才能够模仿这里的行为

**Comment by sybwjdnr** [comment]
更新：

monitor为，validation数据集上每个输出都单独算一个AUC，然后把这些AUC取平均后，监控这个值是否上升，如果上升就更新epoch_best为当前epoch


        valid_loss, valid_acc, valid_cor = convnet:test(valid_seqs, valid_targets, nil, opt.rc)

——_

    if self.target_type == "binary" then
        -- compute AUC
        local AUCs = torch.Tensor(Ydim)
        local roc_points = {}
        for yi = 1,Ydim do
            -- read Yi from file
            local Yi = Yf:partial({1,batcher.num_seqs},{yi,yi}):squeeze()

            if(Yi:sum() == 0) then
                roc_points[yi] = nil
                AUCs[yi] = 1.0

            else
                -- compute ROC points
                roc_points[yi] = ROC.points(preds[{{},yi}], Yi)

                -- compute AUCs
                AUCs[yi] = ROC.area(roc_points[yi])
            end

            collectgarbage()
        end

        return avg_loss, AUCs, roc_points
